python/src/iceberg/types.py

```diff
@@ -46,6 +46,9 @@ def __repr__(self):
     def __str__(self):
         return self._type_string
 
+    def __eq__(self, other):
```

Comment:
Are these objects intended to be immutable?  is there any reason to implement __hash__ for them?

Reply:
Yes, these should be immutable. Is it possible to make them singletons, maybe by hijacking `__new__`? I'm not sure what is normal here.

Reply:
We can do that for some but not for types, e.g. FixedType.

Reply:
Yes, I think we should do that for most cases rather than overriding `__eq__`.

Reply:
@rdblue Thanks for drawing me to this. I think using `__new__` is a great idea you've mentioned a few times elsewhere. It also solved issues with the generics I was trying to make because the syntax is very vague especially for nested fields without any named arguments. I've been working on something since we spoke I've had sitting this morning. @jun-he this seems quite similar https://github.com/apache/iceberg/pull/4016. would definitely be interested in collaborating on this

Reply:
@CircArgs Yeah, definitely. If you plan to make separate new instead of a single one, we can split the work into 2 parts, one for non parameterized one (this PR) and the other for the parameterized one (#4016). What do you think?

Reply:
@jun-he I did end up doing separate `__new__` for all parameterized types. I put a `__new__` on my `PrimitiveType` but putting it on `Type` would work equally well I think. Do you intend to keep your `Singleton` if you can put the `__new__` on `Type`?

Reply:
@CircArgs wondering if it is better to keep Singleton out so any classes can extend it to be singleton? l am also OK to move it into the `__new__` as well. What do you think?

Reply:
@jun-he the Singleton way would make sense to me if it had use beyond types I guess. If the functionality is limited to types I don't know if I can see the difference in functionality since you'll be inheriting from it and rather than hitting the `__new__` on the first parent it will hit it on the second in the case of `class ...(Type, Singleton):...`